service/src/ConstConfigIOGroup.cpp

```diff
@@ -0,0 +1,451 @@
+    bool ConstConfigIOGroup::is_valid_control(const std::string &control_name) const
+    {
+        return false;
+    }
+
+    int ConstConfigIOGroup::signal_domain_type(const std::string &signal_name) const
+    {
+        int result = GEOPM_DOMAIN_INVALID;
+        auto it = m_signal_available.find(signal_name);
+
+        if (it != m_signal_available.end())
+            result = it->second->domain;
+
+        return result;
+    }
+
+    int ConstConfigIOGroup::control_domain_type(const std::string &control_name) const
+    {
+        return GEOPM_DOMAIN_INVALID;
+    }
+
+    int ConstConfigIOGroup::push_signal(const std::string &signal_name,
+                                        int domain_type, int domain_idx)
+    {
+        auto it = m_signal_available.find(signal_name);
+        if (it == m_signal_available.end()) {
+            throw Exception("ConstConfigIOGroup::push_signal(): " +
+                            signal_name + " not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_type != it->second->domain) {
+            throw Exception("ConstConfigIOGroup::push_signal(): domain_type " +
+                            std::to_string(domain_type) +
+                            " not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_idx < 0 ||
+            static_cast<std::size_t>(domain_idx) >= it->second->values.size()) {
+            throw Exception("ConstConfigIOGroup::push_signal(): domain_idx " +
+                            std::to_string(domain_idx) + " out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        for (std::size_t i = 0; i < m_pushed_signals.size(); i++) {
+            const m_signal_ref_s &signal = m_pushed_signals[i];
+            if (it->second == signal.signal_info &&
+                domain_idx == signal.domain_idx) {
+                return static_cast<int>(i);
+            }
+        }
+
+        m_pushed_signals.push_back({.signal_info = it->second,
+                                    .domain_idx = domain_idx});
+        
+        return m_pushed_signals.size() - 1;
+    }
+
+    int ConstConfigIOGroup::push_control(const std::string &control_name,
+                                         int domain_type, int domain_idx)
+    {
+        throw Exception("ConstConfigIOGroup::push_control(): there are no "
+                        "controls supported by the ConstConfigIOGroup",
+                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+    }
+
+    void ConstConfigIOGroup::read_batch(void)
+    {
+    }
+
+    void ConstConfigIOGroup::write_batch(void)
+    {
+    }
+
+    double ConstConfigIOGroup::sample(int batch_idx)
+    {
+        double result = NAN;
+        
+        if (batch_idx < 0 ||
+            static_cast<std::size_t>(batch_idx) >= m_pushed_signals.size()) {
```

Comment:
```suggestion
            batch_idx >= static_cast<int>(m_pushed_signals.size())) {
```
AFAICT, all the other IOGroups do it this way.